docs/dev/Testing.md

```diff
@@ -2,27 +2,108 @@
 
 Drill makes extensive use of [JUnit](http://junit.org/junit4/) and other libraries for testing. This page provides pointers to the information you need to work with Drill tests. We don't repeat that information; you will want to follow the links and read the original material to get a complete understanding of the libraries that Drill uses.
 
-Caveat: information here about Drill is "reverse engineered" from the code; this page has not yet had the benefit of insight from the developers who created Drill's test structure.
-
-# Topics
-
-"Classic" Drill testing techniques
+# Writing Tests
```

Comment:
I've not read the full, original material. This stuff is really quite good.

The one suggestion I'd make is to provide some overall design hints. For example, when should someone use `ClusterFixture` to create an integration test?

When should they use `OperatorFixture` to do a lower-level test?

The answer I followed was that any detailed testing of an operator or other internal mechanism can only be done at the unit level. (All the `RowSet`, `ExternalSort` and `ResultSetLoader` stuff are examples.) This may mean designing the code so it can be tested by controlling dependencies.

Use the `RowSet` mechanism to set up a complete set of input data sets: types, cardinalities, nesting levels: whatever is relevant. As shown in the sort tests, clever coding can automate much of this stuff. There is even a function to generate data for every data type given an int so that can populate types with data that is easy to compare.

Then, once the detailed internal tests are done, use integration tests to ensure the whole system works.

At present, for most readers, integration tests are really the only approach because of the dependencies that Scan has. The in-flight work in the "batch sizing" stuff solves this problem, but only for newer readers designed to use that framework.

The general rule is: test as close to your code as possible, refactoring and breaking dependencies where needed to accomplish this.

On the other hand, if the test is more of a SQL or planner-level concept, then testing at the query level might be fine.

Reply:
Adapted your description here into a Testing Guidelines section of Testing.md